src/web_analysis/test_robots.py

```diff
@@ -0,0 +1,1087 @@
+            "File types should be accessible, but broader disallowance isn't overridden as expected.",
+        )
+
+    def test_implicit_allow(self):
+        """Tests the handling of implicit allow when no explicit rules are set."""
+        agent_rules = {}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "none", "GoogleBot": "none"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Agents should implicitly have complete access if no rules are set.",
+        )
+
+    def test_path_depth_handling(self):
+        """Tests rule application accuracy beyond specified path depth."""
+        agent_rules = {"*": {"Disallow": ["/dir/"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Disallowed directory should extend to all subdirectories and files, but it doesn't.",
+        )
+
+    def test_case_sensitivity_of_paths(self):
+        """Tests the case sensitivity of path rules."""
+        agent_rules = {"*": {"Disallow": ["/Dir/"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Path case sensitivity isn't handled properly, leading to incorrect access rules.",
+        )
+
+    def test_complex_query_parameters(self):
+        """Tests correct handling of URLs with complex query parameters."""
+        agent_rules = {"*": {"Disallow": ["/search", "/search?query=allowed"]}}
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Query parameters should be respected in disallow rules, but aren't being parsed correctly.",
+        )
+
+    def test_overlap_between_allow_and_disallow(self):
+        """Tests the rule precedence when allow and disallow paths overlap."""
+        agent_rules = {
+            "*": {"Disallow": ["/directory/"], "Allow": ["/directory/subdirectory/"]}
+        }
+        all_agents = ["GoogleBot"]
+        expected = {"*": "some", "GoogleBot": "some"}
+        self.assertEqual(
+            interpret_robots(agent_rules, all_agents),
+            expected,
+            "Allow should override Disallow for overlapping paths, but it doesn't.",
+        )
+
+    def test_rule_priority_across_different_user_agents(self):
+        """Ensures specific user-agent rules take precedence over generic ones."""
+        agent_rules = {
+            "*": {"Disallow": ["/private/"]},
+            "GoogleBot": {"Allow": ["/private/stats"]},
+        }
+        all_agents = ["GoogleBot", "BingBot"]
+        expected = {"*": "some", "GoogleBot": "some", "BingBot": "some"}
```

Comment:
GoogleBot should be `all` here, according to https://developers.google.com/search/docs/crawling-indexing/robots/robots_txt#order-of-precedence-for-group-member-records I believe

Reply:
Yep you are right updating this.